app/uk/gov/hmrc/gform/gformbackend/GformConnector.scala

```diff
@@ -85,6 +85,38 @@ class GformConnector(ws: WSHttp, baseUrl: String) {
   def maybeForm(formIdData: FormIdData, formTemplate: FormTemplate)(implicit
     hc: HeaderCarrier,
     ec: ExecutionContext
+  ): Future[Option[Form]] =
+    findForm(formIdData, formTemplate).recoverWith {
+      case UpstreamErrorResponse.WithStatusCode(statusCode, _) if statusCode == StatusCodes.NotFound.intValue =>
+        def getFormByLegacyIds(legacyIds: List[FormTemplateId]): Future[Option[Form]] =
+          legacyIds match {
+            case Nil => Future.successful(None)
+            case legacyFormId :: tail =>
+              getFormTemplate(legacyFormId)
+                .flatMap { legacyFormTemplate =>
+                  val legacyFormIdData = formIdData.withTemplateId(legacyFormTemplate)
+                  logger.info(
+                    s"Attempt to access form $formIdData, but form not found in MongoDB, attempting to look for $legacyFormIdData as a fallback."
+                  )
+                  findForm(legacyFormIdData, legacyFormTemplate).flatMap {
+                    case None    => getFormByLegacyIds(tail)
+                    case Some(_) => createFormFromLegacy(legacyFormIdData, formIdData).map(Some(_))
+                  }
+                }
+                .recoverWith {
+                  case UpstreamErrorResponse.WithStatusCode(statusCode, _)
+                      if statusCode == StatusCodes.NotFound.intValue =>
+                    getFormByLegacyIds(tail)
+                }
+          }
+
+        val legacyFormIds = formTemplate.legacyFormIds.map(_.toList).getOrElse(Nil)
```

Comment:
You've decided to make `legacyFormIds` of type `Option[NonEmptyList[FormTemplateId]]` which is great, but as soon as you can use that representation to your advantage you are throwing it away by converting it to `List[FormTemplateId]`?

Why not to use `legacyFormIds.fold(nothing-to-do-here)(do-the-logic-here)` pattern instead and then `getFormByLegacyIds` can become:

```
def getFormByLegacyIds(legacyIds: NonEmptyList[FormTemplateId]): Future[Option[Form]]
```